File: Minimum Falling path/Memoization.java

import java.util.Arrays;


class Memoization {

    public static int MinimumFallingSum(int[][] arr,int r,int c,int size,int[][] dp) {
        if(r == size-1) return arr[r][c];
        if(dp[r][c] != -1) return dp[r][c];

        int result = Integer.MAX_VALUE;
        for(int i = 0; i < size; i ++) {
            if( i != c)
                result =  Math.min(result, arr[r][c] + MinimumFallingSum(arr, r+1, i, size,dp));
        }

        return dp[r][c] =  result;
    }
    public static void main(String[] args) {
        int[][] arr = {
            {1,2,3},
            {1,2,3},
            {1,2,3}
        };

        int size = arr.length;

        int[][] dp = new int[size][size];

        for(int i = 0 ; i < size; i++)
            Arrays.fill(dp[i],-1);

        int maximum = Integer.MAX_VALUE;
        for(int i = 0 ; i < size; i ++) {
            maximum = Math.min(maximum, MinimumFallingSum(arr,0,i,size,dp));
        }

        System.out.println(maximum);
    }
}

File: Minimum Falling path/Memoization.py
def minimumPath(arr,r,c,size,dp):
    if r == size-1: return arr[r][c]
    if dp[r][c] != -1: return dp[r][c]

    res = float('inf')
    for last in range(size):
        if last !=  c:
            res = min(res,arr[r][c] + minimumPath(arr,r+1,last,size,dp))
    
    dp[r][c] = res
    return dp[r][c]

def minimumFallingSum(arr,size):
    ans = float('inf')

    dp = [ [-1 for _ in range(size)] for _ in range(size)]
    

    for i in range(size):
        ans =  min(ans,minimumPath(arr,0,i,size,dp))
    return ans 

if __name__ == '__main__':
    arr = [
        [1,2,3],
        [1,2,3],
        [1,2,3]
    ]
    size =  len(arr)

    print(minimumFallingSum(arr,size))